docs/proposals/unified-backend-architecture/tech-spec.md

```diff
@@ -0,0 +1,239 @@
+
+To achieve this goal the node must implement robust concurrent model. More specifically, all requests to the wallet
+web API endpoints must be handled concurrently.
+
+**TODO**: To think about the slowest endpoints, like `/api/v1/transactions`: is it possible to handle them
+fully concurrently? I mean publishing to the network, storing in the `wallet-db`, etc.
+
+_Alfredo said: AFAIK that's possible, `acid-state` supports atomic transactions and all the rest of it. Provided we
+write the right `Update` transactions we should be fine._
+
+## Scalability
+
+Initially clusters of nodes will be hosted by IOHK (on AWS EC2). And we need a mechanism of auto-scaling of the
+cluster: it must be able to grow proportionally to growing number of clients.
+
+**TODO**: To think about auto-scale supervisor. Suppose we have 1 cluster that consists of 100 nodes and each of node
+can reliably serve 100 clients. If supervisor sees that number of clients connected to this cluster is growing (and
+becomes more than 10000), it launches additional nodes.
+
+_Domen said: We might not base our autoscaling on number of users, but another resource. It's hard to say how it
+would scale before we run some benchmarks. It could be that bottleneck is CPU and you can't correlate that with number
+of subscribers._
+
+**TODO 2**: Launching additional node is relatively expensive action (in terms of time and memory). So it is better
+if one single node will be able to serve more clients than just increase a number of nodes in the cluster.
+
+## User Stories
+
+These are typical user stories (workflows), with description of API calls.
+
+### Initial State
+
+After installation of mobile client application user does not see any wallets, and he must create at least one to
+work with a client application.
+
+During the first connection with the node the new session is establishing, please read an explanation below.
+
+### Wallet Creation
+
+User creates a new wallet. It produces a `POST`-request to `/api/v1/wallets` (with information about client's public
+key), node creates a new wallet in `wallet-db` and sends response with information about new wallet.
+
+Even if node `N` which serves the client `C` will die, another node `M` (the client `C` will reconnect to) will be able
+to handle information about `C`'s wallet(s), because `wallet-db` is storing in the cluster.
+
+### Wallet Import/Export
+
+Wallet can be imported using client's secret key.
+
+**TODO**: To clarify this process. User can be able to obtain its secret key.
+
+### Wallet Restoring
+
+Wallet can be restored. Use case: user changed a smartphone, install mobile Daedalus client and restore its wallet
+using backup phrase (for example, 12 words) generated during this wallet creation.
+
+Wallet restoring produces a `POST`-request to `/api/v1/wallets/?`, and node checks if such a wallet exists. If so,
+response contains an information about the wallet.
+
+**TODO**: To clarify this process.
```

Comment:
Consult with delegation & restoration design options for details of recovery on backend without the backend having secret keys. Can either choose scheme with hardened or non-hardened keys. It's an interesting trade-off.